dev/Interop/StoragePickers/FileSavePicker.cpp

```diff
@@ -0,0 +1,176 @@
+
+    }
+
+    winrt::Windows::Foundation::IAsyncOperation<winrt::Microsoft::Windows::Storage::Pickers::PickFileResult> FileSavePicker::PickSaveFileAsync()
+    {
+        auto logTelemetry{ StoragePickersTelemetry::FileSavePickerPickSingleFile::Start(m_telemetryHelper) };
+
+        PickerCommon::PickerParameters parameters{};
+        CaptureParameters(parameters);
+        auto defaultFileExtension = m_defaultFileExtension;
+        auto suggestedSaveFile = m_suggestedSaveFile;
+        auto suggestedFileName = m_suggestedFileName;
+        auto fileTypeChoices = m_fileTypeChoices;
+
+        co_await winrt::resume_background();
+        auto cancellationToken = co_await winrt::get_cancellation_token();
+        if (cancellationToken())
+        {
+            logTelemetry.Stop(m_telemetryHelper, false);
+            co_return nullptr;
+        }
+
+        auto dialog = create_instance<IFileSaveDialog>(CLSID_FileSaveDialog, CONTEXT_ALL);
+        parameters.ConfigureDialog(dialog);
+
+        if (!PickerCommon::IsHStringNullOrEmpty(defaultFileExtension))
+        {
+            check_hresult(dialog->SetDefaultExtension(defaultFileExtension.c_str()));
+        }
+
+        if (!PickerCommon::IsHStringNullOrEmpty(suggestedFileName))
+        {
+            check_hresult(dialog->SetFileName(suggestedFileName.c_str()));
+        }
+
+        if (suggestedSaveFile != nullptr)
+        {
+            winrt::com_ptr<IShellItem> shellItem;
+            check_hresult(SHCreateItemFromParsingName(suggestedSaveFile.Path().c_str(), nullptr, IID_PPV_ARGS(shellItem.put())));
+            check_hresult(dialog->SetSaveAsItem(shellItem.get()));
+        }
+
+        {
+            auto hr = dialog->Show(parameters.HWnd);
+            if (FAILED(hr))
+            {
+                logTelemetry.Stop(m_telemetryHelper, false);
+                co_return nullptr;
+            }
+        }
+
+        winrt::com_ptr<IShellItem> shellItem{};
+        check_hresult(dialog->GetResult(shellItem.put()));
+
+        // Get the file path from the dialog
+        wil::unique_cotaskmem_string filePath{};
+        check_hresult(shellItem->GetDisplayName(SIGDN_FILESYSPATH, filePath.put()));
+        winrt::hstring pathStr(filePath.get());
+
+        wil::unique_cotaskmem_string fileName;
+        check_hresult(shellItem->GetDisplayName(SIGDN_NORMALDISPLAY, fileName.put()));
+        std::wstring fileNameStr(fileName.get());
+
+        // Check if the file name has an extension
+        if (fileNameStr.find_last_of(L".") == std::wstring::npos)
+        {
+            // If the user defined file name doesn't have an extension,
+            //     automatically use the first extension from the first category in fileTypeChoices.
+            auto firstCategory = fileTypeChoices.First().Current();
+            auto firstExtension = firstCategory.Value().GetAt(0);
+            pathStr = pathStr + firstExtension;
```

Comment:
>firstExtension [](http://example.com/codeflow?start=32&length=14)

Is there validation anywhere, perhaps in the base IFileDialog code, that the extension starts with a '.'?

Reply:
Thank you very much for your question! This guides me to do some deeper investigation of UWP and our current experimental APIs behavior, and found out that there should be some refinement on our spec/implementation and we could resolve some existing limitations of UWP pickers. We'll continue improve this in following updates.